### PR TITLE
Minor: Simplify `date_trunc` code and add comments

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Follow on to https://github.com/apache/arrow-datafusion/pull/6654 from @Weijun-H  (❤️ )

# Rationale for this change

While reviewing https://github.com/apache/arrow-datafusion/pull/6654 and trying to answer @viirya 's question here https://github.com/apache/arrow-datafusion/pull/6654#discussion_r1244324968

# What changes are included in this PR?
1. Rename `date_trunc_single` to `date_trunc_coarse` to better explain what it does
2. Add comments to explain how it works
3. Remove the use of indirection and call the function directly

# Are these changes tested?
Covered by the (very) nice existing tests added in https://github.com/apache/arrow-datafusion/pull/6654

# Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 1]
```diff
@@ -284,9 +288,7 @@ fn _date_trunc(
     };
 
     // convert to nanoseconds
-    let Some(nano) = (f)(Some(value * scale))? else {
-        return Ok(None);
-    };
+    let nano = date_trunc_coarse(granularity, scale * value)?;
```
Comment (alamb) [inline datafusion/physical-expr/src/datetime_expressions.rs:291]: There is no need for a closure, and the function can be called directly which I think makes things easier to understand

**Review by alamb — COMMENTED** [review]
cc @viirya and @Weijun-H

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 2]
```diff
@@ -214,7 +214,11 @@ fn quarter_month(date: &NaiveDateTime) -> u32 {
     1 + 3 * ((date.month() - 1) / 3)
 }
 
-fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
+/// Tuncates the single `value`, expressed in nanoseconds since the
```
Comment (alamb) [inline datafusion/physical-expr/src/datetime_expressions.rs:217]: I think this is correct -- @Weijun-H  does it match your understanding?

**Review by alamb — COMMENTED** [review]
The `date_trunc` code could be quite a bit faster if it did not check the granularity on each row but instead did that once per column. That can be an optimization for some later time

File: datafusion/physical-expr/src/datetime_expressions.rs (hunk 2 above)
Comment (Weijun-H) [reply datafusion/physical-expr/src/datetime_expressions.rs:217]: Nice simplification 👍 !

**Pull request merged**